src/source.cpp

```diff
@@ -378,6 +410,96 @@ CompiledSourceWrapper::~CompiledSourceWrapper()
 #endif
 }
 
+//==============================================================================
+// MeshSource implementation
+//==============================================================================
+
+MeshSource::MeshSource(pugi::xml_node node)
+{
+  // Check for particle type
+  if (check_for_node(node, "particle")) {
+    auto temp_str = get_node_value(node, "particle", true, true);
+    if (temp_str == "neutron") {
+      particle_ = ParticleType::neutron;
+    } else if (temp_str == "photon") {
+      particle_ = ParticleType::photon;
+      settings::photon_transport = true;
+    } else {
+      fatal_error(std::string("Unknown source particle type: ") + temp_str);
+    }
+  }
+
+  int32_t mesh_id = stoi(get_node_value(node, "mesh"));
+  int32_t mesh_idx = model::mesh_map.at(mesh_id);
+  const auto& mesh = model::meshes[mesh_idx];
+
+  std::vector<double> strengths;
+  // read all source distributions and populate strengths vector for MeshSpatial
+  // object
+  for (auto source_node : node.children("source")) {
+    sources_.emplace_back(IndependentSource(source_node));
+    strengths.push_back(sources_.back().strength());
+  }
+
+  // the number of source distributions should either be one or equal to the
+  // number of mesh elements
+  if (sources_.size() > 1 && sources_.size() != mesh->n_bins()) {
+    fatal_error(fmt::format("Incorrect number of source distributions ({}) for "
+                            "mesh source with {} elements.",
+      sources_.size(), mesh->n_bins()));
+  }
+
+  space_ = std::make_unique<MeshSpatial>(mesh_idx, strengths);
+}
+
+SourceSite MeshSource::sample(uint64_t* seed) const
+{
+  SourceSite site;
+  site.particle = particle_;
+
+  int n_reject = 0;
+  static int n_accept = 0;
+
+  // sample location and element from mesh
+  auto mesh_location = space_->sample_mesh(seed);
+
+  site.r = mesh_location.second;
+
+  int32_t element = mesh_location.first;
+
+  const auto& src = source(element);
+
+  site.u = src.angle()->sample(seed);
+
+  Distribution* e_dist = src.energy();
```

Comment:
Rather than sampling each of the individual distributions from the independent source, we could just fully sample a source site from the underlying distribution and then replace the sample spatial point with the one from the mesh.